### PR TITLE
Minor additions to IntegratePeakSkew

### PR DESCRIPTION
**Pull request by RichardWaiteSTFC**
**Description of work.**

Bug Fixes:

- Setting colorfill ` vmin > vmax` for mostly empty arrays
-  Close figure if `OSError` while trying to write to file
- Fix bug that didn't set HKL of peak if `UpdatePeakPosition = True`
- Reset progress bar to account for time spent plotting peaks

Improvements:

- Speed-up plotting by re-using artists of colorfill and 1D plot
- Use a robust (to outliers) means of estimating resolution parameters

New features:
- Add option to specify min number of TOF bins in valid peak
- Add option to scale dth by wavelength in TOF resolution (for instruments with beam divergence proportional to wavelength)
image

**To test:**

To-Do

*There is no associated issue.*

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by DannyHindson** [comment]
@RichardWaiteSTFC - is this ready for review now?

**Comment by RichardWaiteSTFC** [comment]
> @RichardWaiteSTFC - is this ready for review now?

Thanks for checking, just got one more small change to go!

**Review by DannyHindson — CHANGES_REQUESTED** [review]
Looks good. Just a few minor things in the detail comments. Also:
- perhaps it's worth adding another couple of peaks to the SXD docs usage example so that the new TOF resolution plot has more than one point?
- there's a .gitkeep file in the file list. Please can you remove?
- as a general rule we should only add new algorithm parameters at the end of the list but since this algorithm is very new\in evaluation I think it's fine here to add them in the middle to keep the grouping sensible
- Just for completeness I'll note here that there's also a change to make Lorentz correction optional that's not in the PR description but that looks fine

File: docs/source/algorithms/IntegratePeaksSkew-v1.rst [hunk 1]
```diff
@@ -151,6 +155,11 @@ showing the peak mask and the data integrated over the TOF window that maximises
 focused, background subtracted spectra with vertical lines to mark the initial and optimal TOF window and the peak
 centre.
 
+.. figure:: ../images/EstimatedResolutionParametersSkewIntegration.png
+    :align: center
+    :width: 50%
+    :alt: (Left) Found fractional TOF window and estimated curves at 4 different wavelengths from linear fit shown on
+          (Right)
```
[the remark this answers is not in the document]
Comment (RichardWaiteSTFC) [reply docs/source/algorithms/IntegratePeaksSkew-v1.rst:162]: I could put it where I discuss the resolution (i.e. step 1) near the top?

[the remark this answers is not in the document]
Comment (DannyHindson) [reply docs/source/algorithms/IntegratePeaksSkew-v1.rst:162]: Would that interrupt the flow of the step by step description a bit? Depends on how big it is and how it looks. Also fine just to leave it where it is and just mention the new plot in the text under the "Plotting" heading

File: Framework/PythonInterface/plugins/algorithms/IntegratePeaksSkew.py [hunk 2]
```diff
@@ -539,6 +594,11 @@ def PyInit(self):
                              doc="dTheta resolution (estimated from width at forward scattering minus "
```
[the remark this answers is not in the document]
Comment (RichardWaiteSTFC) [reply Framework/PythonInterface/plugins/algorithms/IntegratePeaksSkew.py:594]: Yes it's radians, I will clarify the docstring and in the plot

File: Framework/PythonInterface/plugins/algorithms/IntegratePeaksSkew.py [hunk 3]
```diff
@@ -539,6 +594,11 @@ def PyInit(self):
...
                                  "contribution from moderator, dT0/T0, and path length dL/L).")
         self.setPropertySettings("ThetaWidth", condition_to_use_resn)
+        self.declareProperty(name="ScaleThetaWidthByWavelength", defaultValue=False, direction=Direction.Input,
```
[the remark this answers is not in the document]
Comment (DannyHindson) [reply Framework/PythonInterface/plugins/algorithms/IntegratePeaksSkew.py:597]: Yes sure. Just to check the algorithm still completes successfully with this set to True

File: Framework/PythonInterface/plugins/algorithms/IntegratePeaksSkew.py [hunk 4]
```diff
@@ -740,41 +809,119 @@ def PyExec(self):
...
+        scaled_cot_th_sq = ((wavelengths ** 2) / np.tan(thetas)) ** 2 if scale_dth else (1 / np.tan(thetas))
+        frac_tof_widths = np.array([pk_data.get_dTOF_over_TOF() for pk_data in peak_data_collection
+                                    if pk_data.status == PEAK_MASK_STATUS.VALID])
+        estimated_dt0_over_t0, estimated_dth = -1, -1
+        if len(thetas) > 1:
+            # robust estimation of params for (dT/T)^2 = slope*cot(theta)^2 + intercept
+            # if scale_dth cot(th) -> wl*cot(th)
+            slope, intercept = self.estimate_linear_params(scaled_cot_th_sq, frac_tof_widths ** 2)
+            estimated_dt0_over_t0 = np.sqrt(intercept)
+            estimated_dth = np.sqrt(slope)
```
[the remark this answers is not in the document]
Comment (RichardWaiteSTFC) [reply Framework/PythonInterface/plugins/algorithms/IntegratePeaksSkew.py:827]: Yes good catch they should indeed be inside the `if` on the next line!

File: Framework/PythonInterface/plugins/algorithms/IntegratePeaksSkew.py [hunk 5]
```diff
@@ -740,41 +809,119 @@ def PyExec(self):
...
+                colors = line.get_cmap().colors
+                nwl = 4
+                wls = np.linspace(wavelengths.min(), wavelengths.max(), nwl)
+                icolors = np.linspace(0, len(colors)-1, nwl, dtype=int)
+                for wl, icolor in zip(wls, icolors):
+                    ax[0].plot(np.degrees(xvals), np.sqrt(intercept * ((wl**2) / np.tan(xvals) ** 2) + slope),
+                               '-', color=colors[icolor])
+                ax[0].set_ylim(*ylim)  # reset limits so suitable for data point coverage not fit lines
+            # add resolution parameters to the plot title
+            ax[1].set_title(rf"$d\theta$={estimated_dth:.2E}" + "\n$dT_{bk}/T_{bk}$" + f"={estimated_dt0_over_t0:.2E}")
```
[the remark this answers is not in the document]
Comment (RichardWaiteSTFC) [reply Framework/PythonInterface/plugins/algorithms/IntegratePeaksSkew.py:923]: I was debating with myself whether to have the input `ThetaWidth` in degrees instead, that way the units on the plot can be consistent and the number might be a bit more meaningful to the user?

**Review by DannyHindson — APPROVED** [review]
Thanks for the changes. All looks good now

**Pull request merged**